Remove OTP requirement for enabling Webauthn as a second factor

nicolastemciuc opened the pull request:
#### Summary

This PR removes the OTP requirement when enabling WebAuthn as a second factor, simplifying the 2FA setup flow for users.

It builds upon the work initially done by @santiagorodriguez96 in https://github.com/mastodon/mastodon/pull/25087.

Fixes https://github.com/mastodon/mastodon/issues/16693.

nicolastemciuc committed:
feat: allow adding WebAuthn security keys without OTP enabled

Co-authored-by: Santiago Rodriguez <santiago.rodriguez@cedarcode.com>

nicolastemciuc committed:
feat: redirect to two factor methods index after disabling 2FA

Co-authored-by: Santiago Rodriguez <santiago.rodriguez@cedarcode.com>

nicolastemciuc committed:
test: update specs to not require OTP for enabling WebAuthn as 2FA

Co-authored-by: Santiago Rodriguez <santiago.rodriguez@cedarcode.com>

nicolastemciuc committed:
feat: allow users to remove authenticator app from 2FA methods

Co-authored-by: Santiago Rodriguez <santiago.rodriguez@cedarcode.com>

nicolastemciuc committed:
Merge branch 'main' into nt--remove-otp-requirement-for-enabling-webauthn

nicolastemciuc committed:
Merge remote-tracking branch 'origin/master' into nt--remove-otp-requirement-for-enabling-webauthn

nicolastemciuc committed:
Merge branch 'main' into nt--remove-otp-requirement-for-enabling-webauthn

ClearlyClaire reviewed (commented):
Thanks for your contribution!

It appears very reasonable, however the log-in flow has not been fully adapted and still offers to connect using TOTP or using a backup code when TOTP is not enabled.

(By the way, should backup codes be generated when Webauthn is set?)

trwnh commented:
to rephrase:

if i want to use 3 factors, i also can't do that currently. i can only do 1 (a password or a backup code), or 2 (one of which must be a password). if i want to avoid passwords or be more secure than 2fa, then i am out of luck.

i agree that removing the TOTP requirement for WebAuthn keys is a good first step. i'm just advocating for going beyond that in the future and letting people move away from passwords when they want to.

wrt backup codes i can see an argument for not always generating them, but the user can "simply" not save them anywhere, i guess? that leaves them open to potential brute-forcing of backup codes, though.

santiagorodriguez96 committed:
Merge branch 'main' into nt--remove-otp-requirement-for-enabling-webauthn

santiagorodriguez96 committed:
fix(OAuth): redirect missing-2FA users to TOTP setup page instead of methods index

When a user's role requires 2FA but they haven't set it up yet,
`require_functional!` redirects them via `mfa_setup_path`. Previously
this pointed to the methods index, which in turn redirected to OTP
setup via a `before_action`. After removing that guard to allow the
methods index to be accessible without 2FA, the redirect chain broke.
Point `mfa_setup_path` directly to the OTP setup page to restore the
original OAuth authorization flow behavior.

santiagorodriguez96 committed:
feat: require password challenge before adding WebAuthn security keys

santiagorodriguez96 committed:
feat: create backup codes after enabling 2FA by adding a security key

santiagorodriguez96 committed:
feature: delete backup codes if last WebAuthn credential is removed and OTP is not enabled

santiagorodriguez96 committed:
tests: add controller specs

ClearlyClaire reviewed (commented):
Thank you for your contribution!

It appears to mostly work as expected, but there are a few quirks:
- passing the recovery codes as HTML to a bit of javascript does work, but I'm not very comfortable with that
- there seem to be some inconsistency in the way OTP backup codes are cleared: if i add both TOTP and a security key then remove them, the OTP backup codes are not cleared, and thus not generated again when I add a security key again
- the role requirement for 2FA redirects to TOTP setup only, this is not a big issue but probably worth addressing

santiagorodriguez96 committed:
Merge branch 'main' into nt--remove-otp-requirement-for-enabling-webauthn

santiagorodriguez96 committed:
chore: remove translated strings

In order to avoid unnecessary noise to the PR. Translations will be
added in the next Crowdin merge.

References: #37801

santiagorodriguez96 commented on app/controllers/settings/two_factor_authentication/webauthn_credentials_controller.rb:
```diff
@@ -43,23 +45,46 @@ def create
+
+            generated_otp_backup_codes = false
+            if current_user.otp_backup_codes.blank?
+              current_user.otp_secret = User.generate_otp_secret(32)
```
Yes, I was essentially following what's done in the 2FA confirmations controller:

https://github.com/cedarcode/mastodon/blob/e563a99723db7cbcd8744c42818b94a9370fbfd3/app/controllers/settings/two_factor_authentication/confirmations_controller.rb#L25

santiagorodriguez96 commented on app/controllers/settings/two_factor_authentication/webauthn_credentials_controller.rb:
```diff
@@ -43,23 +45,46 @@ def create
+                    locals: { recovery_codes: recovery_codes },
+                    formats: :html,
+                    layout: false
+                  ),
```
To be honest I totally agree with you. I made it like this because we don't have any endpoint to show the recovery codes to redirect to – currently recovery codes are only show once by simply rendering a view when setting up TOTP – so I preferred the simplicity of a similar approach rather than creating an endpoint which IMO is a different new feature that we should give some thought to it (for instance, if it's okay for users to view their recovery codes through the UI; we should probably prevent their access with a password; and such) and could essentially change that philosophy of only being able of viewing the recovery codes once.

But if you are okay with it, perhaps we could simply add the endpoint (not accessible via the UI yet) behind a password challenge and redirect to it if you are okay with it.

santiagorodriguez96 replied:
Hi @ClearlyClaire!

I've refactored this by adding the `recovery_codes#index` endpoint that reads the codes from the session – which is populated with them by `webauthn_credentials#create`. Take a look at db4404f08687eeaa853301e733e204df8ec5b09d to see the changes that relates to this refactor. Let me know what you think 🙂

santiagorodriguez96 committed:
feat: button for re-generating recovery codes incorrectly not being shown

The button and hint for re re-generating the recovery codes is not
appearing when TOTP is not enabled, even though recovery codes could be
generated by adding a security key.

santiagorodriguez96 reviewed (commented):
Hey @ClearlyClaire! Thank you so much for your review! Sorry that it took me so long to get back to you 😕 

> * there seem to be some inconsistency in the way OTP backup codes are cleared: if i add both TOTP and a security key then remove them, the OTP backup codes are not cleared, and thus not generated again when I add a security key again

Hmm can you be more specific about which steps are you doing to reproduce this? I'm not being able to reproduce it myself 🤔 

> * the role requirement for 2FA redirects to TOTP setup only, this is not a big issue but probably worth addressing

Are you talking about the OAuth+MFA requirement flow? In that case I'm aware! We could probably add a link below the "Set up" button or straight up changing the flow to redirect to the page for adding a security key altogether (with a TOTP fallback maybe?). Ultimately, it feels to me though that this is a little bit out of scope for this PR (which is pretty big already) so I think it might be for the best to give some though to how we want that flow to look like and leave those changes to a separate effort in a different PR. I can help with that too!

santiagorodriguez96 commented:
> If you do rebase it, please leave out the changes to the translation files in order to make the PR easier to review, thanks!

Done in 860d4e9!

santiagorodriguez96 committed:
test: fix failing recovery code spec

Now that we updated the "Generate Recovery Codes" button in order to be
shown only when the user has their recovery codes generated, we have to
update the setup of the tests in order to reflect the correct state.

santiagorodriguez96 commented:
Hey @ClearlyClaire! The [`check-i18n`](https://github.com/mastodon/mastodon/actions/runs/26780143241/job/79025468738?pr=35811#logs) check is failing given that now, after the changes in 860d4e9, we are not removing the (now unused) `otp_required` translations. Is that expected? Or should I go ahead and remove them?

ClearlyClaire commented:
> Hey @ClearlyClaire! The [`check-i18n`](https://github.com/mastodon/mastodon/actions/runs/26780143241/job/79025468738?pr=35811#logs) check is failing given that now, after the changes in [860d4e9](https://github.com/mastodon/mastodon/commit/860d4e995efe191e62a700c101907b8a4002a639), we are not removing the (now unused) `otp_required` translations. Is that expected? Or should I go ahead and remove them?

Hi! Sorry for the late reply! It appears I was mistaken as to why there was an issue before. This error should go away after https://github.com/mastodon/mastodon/pull/39347 is merged

santiagorodriguez96 committed:
Merge branch 'main' into nt--remove-otp-requirement-for-enabling-webauthn

santiagorodriguez96 committed:
Merge branch 'main' into nt--remove-otp-requirement-for-enabling-webauthn

santiagorodriguez96 committed:
chore(`rubocop`): fix `RSpec/MatchWithSimpleRegex` offenses

gracjankn commented:
Is there any new progress anyone can share on this topic? The last activity here was over two months ago. This is a very needed first step towards better passkey support on Mastodon and I sincerely hope that it'll get merged into v4.7.

renchap commented:
From our latest internal discussions, there are still a few issues outlined above that need to be fixed. The [main one](https://github.com/mastodon/mastodon/pull/35811#discussion_r2947521792) is having to pass a lot of server-generated HTML through JS, this is not very clean and there is likely a better solution here.

gracjankn commented:
Thank you for the update! I’m glad it’s still on the radar! 👍🏻

santiagorodriguez96 commented on app/controllers/settings/two_factor_authentication/recovery_codes_controller.rb:
```diff
@@ -9,6 +9,12 @@ class RecoveryCodesController < BaseController
 
       before_action :require_challenge!
 
+      def index
+        @recovery_codes = session.delete(:new_recovery_codes)
+
+        redirect_to settings_two_factor_authentication_methods_path if @recovery_codes.blank?
+      end
```
This new `index` action inherits `RecoveryCodesController`'s controller-wide `require_challenge!`, but in practice that almost never fires. The only entry point is the redirect from `webauthn_credentials#create`, which lands after the challenge passed on `#new`, so the one-hour grace period should prevent the challenge from being required.

However, as the challenge is requested when `webauthn_credentials#new` is loaded and options and create run whenever the user actually clicks "Add" – and neither is challenged, so neither refreshes `challenge_passed_at`, theoretically a user could leave the page stale and click on "Add" after the challenge grace period. In that case, the challenge will be requested after the ceremony is completed and before showing the recovery codes.

I've left the guard in place since it costs nothing, but happy to scope it to `only: [:create]` if you'd rather not add something that rarely fires.

santiagorodriguez96 commented on app/controllers/settings/two_factor_authentication/otp_authentication_controller.rb:
```diff
@@ -22,7 +22,12 @@ def create
 
       def destroy
-        current_user.disable_otp_login!
+        if current_user.webauthn_enabled?
+          current_user.disable_otp_login!
+        else
+          current_user.disable_two_factor!
+          UserMailer.two_factor_disabled(current_user).deliver_later!
+        end
```
@ClearlyClaire I think this was the issue that you reported some time ago:

> there seem to be some inconsistency in the way OTP backup codes are cleared: if i add both TOTP and a security key then remove them, the OTP backup codes are not cleared, and thus not generated again when I add a security key again

The issue was that if you removed TOTP last, the recovery codes weren't being deleted. Fixed that in a96eebca0bf96ac094b0c6e2755cca976.

santiagorodriguez96 commented on app/controllers/settings/two_factor_authentication/recovery_codes_controller.rb:
```diff
@@ -9,6 +9,12 @@ class RecoveryCodesController < BaseController
+        redirect_to settings_two_factor_authentication_methods_path if @recovery_codes.blank?
+      end
+
       def create
```
I think there's an issue in `master` where a user could leave a stale tab in the two factor authentication methods index page and, in another page, click in "Disable 2FA". After doing that, the user could go to the stale index page and click in "Generate recovery codes" and it will work, causing the user to not have any two factor method enable but recovery codes generated.

santiagorodriguez96 commented on app/controllers/settings/two_factor_authentication/webauthn_credentials_controller.rb:
```diff
@@ -59,7 +69,7 @@ def create
+        render json: { redirect_path: redirect_path }, status: status
       end
 
       def destroy
```
I think we should require re-authentication for removing security keys. Seemed out of scope to me so I left it to tackle it as part of a separate effort.

santiagorodriguez96 commented on app/controllers/settings/two_factor_authentication/otp_authentication_controller.rb:
```diff
@@ -20,11 +21,26 @@ def create
 
+      def destroy
+        if current_user.webauthn_enabled?
+          current_user.disable_otp_login!
```
We should probably send an email in this case, isn't it?
